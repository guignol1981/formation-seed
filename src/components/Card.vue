<template>
    <m-panel
        class="post"
        :highlighted="true"
        :padding="false"
        :shadow="true"
        skin="light"
    >
        <div v-if="title" slot="header" class="post__header">
            <h2 class="post__title">{{ title }}</h2>
        </div>
        <div class="post__body" v-html="body"></div>
        <div slot="footer" class="post__footer">
            <div>
                <slot name="actions-left">
                    actions
                </slot>
            </div>
            <div>
                <slot name="actions-right">
                    actions
                </slot>
            </div>
        </div>
    </m-panel>
</template>

<script lang="ts">
    import Vue from 'vue';
    import Component from 'vue-class-component';
    import { Prop, Emit } from 'vue-property-decorator';

    @Component
    export default class Card extends Vue {
        @Prop({ required: true })
        public body!: string;

        @Prop()
        public title?: string;
    }
</script>

<style lang="scss" scoped>
    @import '~@ulaval/modul-components/dist/styles/commons';

    .post {
        position: relative;

        + .post {
            margin-top: $m-spacing--l;
        }

        &__header {
            display: flex;
            align-items: center;
            font-weight: bold;
            font-size: $m-font-size--h3;
            padding: 1rem;
        }

        &__title {
            margin: 0 0 0 $m-spacing--s;

            @media (min-width: $m-mq--min--s) {
                margin: 0 0 0 $m-spacing--m;
            }
        }

        &__body {
            text-align: start;
            padding: $m-spacing--m;
            min-height: $m-spacing--xl;
        }

        &__footer {
            display: flex;
            justify-content: space-between;
            align-items: center;
            background-color: $m-color--white;
            padding: $m-spacing--s !important;

            & > div > * {
                margin-left: $m-spacing--s;
            }
        }
    }
</style>
