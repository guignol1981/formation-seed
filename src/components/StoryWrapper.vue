<template>
    <div class="m-u--app-body">
        <slot></slot>
    </div>
</template>

<script lang="ts">
    import Vue from 'vue';
    import Component from 'vue-class-component';

    @Component
    export default class StoryWrapper extends Vue {}
</script>

<style lang="scss">
    $m-font-path: '~@ulaval/modul-components/dist/assets/fonts/'; //this is required.
    @import '~@ulaval/modul-components/dist/styles/main.scss';
</style>
