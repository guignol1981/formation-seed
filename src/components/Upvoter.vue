<template>
    <div>
        <m-icon-button
            iconName="m-svg__arrow--up"
            buttonSize="32px"
            iconSize="20px"
            :skin="ownerHasUpvoted ? 'primary' : 'light'"
            @click="onUpvoteClicked()"
        ></m-icon-button>
        {{ votable.upvotes.length }}
        <m-icon-button
            iconName="m-svg__arrow--down"
            buttonSize="32px"
            iconSize="20px"
            :skin="ownerHasDownvoted ? 'primary' : 'light'"
            @click="onDownvoteClicked()"
        ></m-icon-button>
        {{ votable.downvotes.length }}
    </div>
</template>

<script lang="ts">
    export interface Votable {
        upvotes: string[];
        downvotes: string[];
    }

    import Vue from 'vue';
    import Component from 'vue-class-component';
    import { Prop, Emit, Watch } from 'vue-property-decorator';
    const _ = require('lodash');

    @Component
    export default class VueVoter extends Vue {
        @Prop({ required: true })
        public votable!: Votable;

        @Prop({ required: true })
        public ownerId!: string;

        public get ownerHasUpvoted(): boolean {
            return this.votable.upvotes.includes(this.ownerId);
        }

        public get ownerHasDownvoted(): boolean {
            return this.votable.downvotes.includes(this.ownerId);
        }

        public onUpvoteClicked(): void {
            if (this.ownerHasUpvoted) {
                this.votable.upvotes = this.votable.upvotes.filter(
                    u => u !== this.ownerId
                );
                return;
            }

            this.votable.upvotes.push(this.ownerId);
            this.votable.downvotes = this.votable.downvotes.filter(
                d => d !== this.ownerId
            );
        }

        public onDownvoteClicked(): void {
            if (this.ownerHasDownvoted) {
                this.votable.downvotes = this.votable.downvotes.filter(
                    d => d !== this.ownerId
                );
                return;
            }

            this.votable.downvotes.push(this.ownerId);
            this.votable.upvotes = this.votable.upvotes.filter(
                u => u !== this.ownerId
            );
        }
    }
</script>

<style lang="scss" scoped></style>
