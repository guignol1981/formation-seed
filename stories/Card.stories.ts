import { Component } from 'vue';
import Card from '../src/components/Card';
import StoryWrapper from '../src/components/StoryWrapper';
import Upvoter from '../src/components/Upvoter';
import StoryRouter from 'storybook-vue-router';
import { action } from '@storybook/addon-actions';
import { text, withKnobs } from '@storybook/addon-knobs';

export default {
    title: 'Card',
    component: Card,
    decorators: [StoryRouter(), withKnobs]
};

export const Default: () => Component = (): Component => ({
    components: { Card, StoryWrapper },
    props: {
        title: {
            default: text('title', 'title')
        },
        body: {
            default: text('body', 'body')
        }
    },
    template: `
        <StoryWrapper>
                <Card 
                    :title="title"
                    :body="body"
                >
                </Card>
        </StoryWrapper>
    `
});

export const WithUpvoter: () => Component = (): Component => ({
    components: { Card, StoryWrapper, Upvoter },
    data: () => ({
        votable: {
            upvotes: ['3', '2', '3'],
            downvotes: ['1']
        },
        ownerId: '1'
    }),
    template: `
        <StoryWrapper>
                <Card 
                    :title="'title'"
                    :body="'body'"
                    >
                    <Upvoter 
                        slot="actions-left"
                        :votable="votable"
                        :ownerId="ownerId"
                    >
                    </Upvoter>
                </Card>
        </StoryWrapper>
    `
});
