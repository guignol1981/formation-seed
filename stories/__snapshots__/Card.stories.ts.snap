// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Card Default 1`] = `
<div
  class="m-u--app-body"
>
  <article
    class="m-panel post m--is-skin-light m--has-shadow m--has-border m--is-highlighted"
    data-v-74fdcb1f=""
  >
    <header
      class="m-panel__header m--no-padding"
      data-v-74fdcb1f=""
    >
      <div
        class="m-panel__header__content"
        data-v-74fdcb1f=""
      >
        <div
          class="post__header"
          data-v-74fdcb1f=""
        >
          <h2
            class="post__title"
            data-v-74fdcb1f=""
          >
            title
          </h2>
        </div>
      </div>
       
      <!---->
       
      <!---->
    </header>
     
    <div
      class="m-panel__body m--no-padding"
      data-v-74fdcb1f=""
    >
      
    
    
      <div
        class="post__body"
        data-v-74fdcb1f=""
      >
        body
      </div>
       
    </div>
     
    <footer
      class="m-panel__footer m--no-padding"
      data-v-74fdcb1f=""
    >
      <div
        class="post__footer"
        data-v-74fdcb1f=""
      >
        <div
          data-v-74fdcb1f=""
        >
          
                actions
            
        </div>
         
        <div
          data-v-74fdcb1f=""
        >
          
                actions
            
        </div>
      </div>
    </footer>
  </article>
</div>
`;

exports[`Storyshots Card With Upvoter 1`] = `
<div
  class="m-u--app-body"
>
  <article
    class="m-panel post m--is-skin-light m--has-shadow m--has-border m--is-highlighted"
    data-v-74fdcb1f=""
  >
    <header
      class="m-panel__header m--no-padding"
      data-v-74fdcb1f=""
    >
      <div
        class="m-panel__header__content"
        data-v-74fdcb1f=""
      >
        <div
          class="post__header"
          data-v-74fdcb1f=""
        >
          <h2
            class="post__title"
            data-v-74fdcb1f=""
          >
            title
          </h2>
        </div>
      </div>
       
      <!---->
       
      <!---->
    </header>
     
    <div
      class="m-panel__body m--no-padding"
      data-v-74fdcb1f=""
    >
      
    
    
      <div
        class="post__body"
        data-v-74fdcb1f=""
      >
        body
      </div>
       
    </div>
     
    <footer
      class="m-panel__footer m--no-padding"
      data-v-74fdcb1f=""
    >
      <div
        class="post__footer"
        data-v-74fdcb1f=""
      >
        <div
          data-v-74fdcb1f=""
        >
          <div
            data-v-74fdcb1f=""
          >
            <button
              class="m-icon-button m--is-skin-light m--has-ripple"
              data-v-4b160bdb=""
              style="width: 32px; height: 32px;"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="m-icon m-icon-button__icon"
                data-v-3d21e689=""
                data-v-4b160bdb=""
                height="20px"
                width="20px"
              >
                <!---->
                 
                <use
                  aria-hidden="true"
                  class=""
                  data-v-3d21e689=""
                  xlink:href="#m-svg__arrow--up"
                />
              </svg>
               
              <!---->
            </button>
            
    3
    
            <button
              class="m-icon-button m--is-skin-primary m--has-ripple"
              data-v-4b160bdb=""
              style="width: 32px; height: 32px;"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="m-icon m-icon-button__icon"
                data-v-3d21e689=""
                data-v-4b160bdb=""
                height="20px"
                width="20px"
              >
                <!---->
                 
                <use
                  aria-hidden="true"
                  class=""
                  data-v-3d21e689=""
                  xlink:href="#m-svg__arrow--down"
                />
              </svg>
               
              <!---->
            </button>
            
    1

          </div>
        </div>
         
        <div
          data-v-74fdcb1f=""
        >
          
                actions
            
        </div>
      </div>
    </footer>
  </article>
</div>
`;
